### ISSUE TITLE
divide by zero using quantity and masked array

### ISSUE THREAD

**Issue by elinor-lev** [astropy/astropy#7365]
When dividing a quantity by a masked array with zeros, this invalid behavior occurs:
```
a = 1e18 * u.m
b = np.ma.array([0.,0])
print a/b

[1. 1.] m
```

whereas the the scalar is actually correct:
```
print a/b[0]

inf m
```

This is not observed if using a simple numpy array:
```
b = np.array([0.,0])
print a/b

[inf inf] m
```

or alternatively if not using quantity but a value, it results in array with masked values:
```
a = 1e18
b = np.ma.array([0.,0])
print a/b

[-- --]
```

not sure if this is a bug, feature, astropy or numpy issue?

I should not that this worked fine until a recent update of either numpy/astropy (not sure which).
I'm using numpy 1.14.2 and astropy 2.0.5

**Comment by pllim** [comment]
Thank you for reporting this. I see this in Python 3 with astropy 3.1.dev21672 as well.
```python
>>> a / b
.../astropy/units/quantity.py:639: RuntimeWarning: divide by zero encountered in true_divide
  result = super().__array_ufunc__(function, method, *arrays, **kwargs)
<Quantity [1., 1.] m>
```

**Comment by elinor-lev** [comment]
This appears to be partly the due numpy masked array behavior of masking inf that result from division by zero (see [#5384](https://github.com/numpy/numpy/issues/5384) )


```python
In [9]: a = 1. 
   ...: b = np.ma.array([0.,0])
   ...: a/b
.../anaconda/bin/ipython:3: RuntimeWarning: divide by zero encountered in divide
  # -*- coding: utf-8 -*-
Out[9]: 
masked_array(data=[--, --],
             mask=[ True,  True],
       fill_value=1e+20,
            dtype=float64)

In [10]: a/b * u.m
.../anaconda/bin/ipython:1: RuntimeWarning: divide by zero encountered in divide
Out[10]: <Quantity [1., 1.] m>

In [11]: a/b.data * u.m
.../anaconda/bin/ipython:1: RuntimeWarning: divide by zero encountered in divide
Out[11]: <Quantity [inf, inf] m>
```



since the inf gets masked, multiplying by the astropy unit wrongly converts the maksed inf value into a '1'.
If instead of using the masked array we use the data, the expected value is returned.

**Comment by mhvk** [comment]
A proper fix on the astropy side will require support for `MaskedQuantity`... (#8133, #1857)

**Comment by mhvk** [comment]
Will be fixed by #11127, although only in the sense that we'll have our own Masked class that does not have this problem.
